### PR TITLE
(QA-3457) Switch to using ABS for AWS provisioning

### PR DESCRIPTION
**Pull request by samwoods1**
Updated hosts files to use ABS hypervisor
Broke out rake tasks into multiple tasks, one to provision, one to execute beaker, one to encapsulate both of those and finally, one to deprovision.
Removed workarounds for AWS that are no longer necessary: 10_etc_hosts.rb, 40_ec2_fixes.rb, 20_use_internal_ips.rb and some of the FOSS install code in perf_helper.rb
Fixed 20_foss_install/10_install_dev_repos.rb - it was missing the call to the helper method

### CODE REVIEW

File: rakefile [hunk 1]
```diff
@@ -1,13 +1,85 @@
...
+  req = Net::HTTP::Post.new(uri, 'Content-Type' => 'application/json', )
+  req.body = {'resources': {'centos-7-x86-64-west': 2},'job': { 'id': "#{ENV['ABS_JOB_ID']}", 'tags': {'user': 'jenkins','jenkins_build_url': 'https://jenkins/job/local-perf-test'}}}.to_json
+  http = Net::HTTP.new(uri.host, uri.port)
+  http.use_ssl = true
+
+  puts "sending request body: #{req.body}"
+  puts "to uri: #{uri}"
+  res = nil
+
+  # Wait up to 5 minutes for a response from ABS with provisioned hosts
```
Comment (er0ck) [inline rakefile]: this could probably be placed in a method, outside the task.  as-is, it's hard for me to follow what the task is doing exactly.

File: rakefile [hunk 2]
```diff
@@ -1,13 +1,85 @@
...
+        sleep(10)
+      end
+    end
+  end
+  ENV['ABS_RESOURCE_HOSTS'] = res.body
+  puts "ABS_RESOURCE_HOSTS=#{ENV['ABS_RESOURCE_HOSTS']}"
+end
+
+rototiller_task :performance_deprovision_with_abs do
+  raise 'performance_deprovision_with_abs task requires values for env variables: ABS_RESOURCE_HOSTS and ABS_JOB_ID' unless ENV['ABS_RESOURCE_HOSTS'] && ENV['ABS_JOB_ID']
```
Comment (er0ck) [inline rakefile]: do we do this currently due to deficiencies in rototiller's required env vars?  there is very likely nothing to handle this exception.

File: rakefile [hunk 3]
```diff
@@ -106,7 +178,7 @@ rototiller_task :performance do |t|
       option.name = '--preserve-hosts'
       option.message = 'Whether to preserve hosts or not'
       option.add_argument do |arg|
-        arg.name = 'always'
+        arg.name = 'never'
```
Comment (er0ck) [inline rakefile]: this is probably useful outside of this commit.  should be separated.  e.g. if the other code needs reverting, we'd want this to stay.  correct?

File: rakefile [hunk 4]
```diff
@@ -146,6 +218,13 @@ rototiller_task :acceptance do |t|
   Rake::Task["performance"].execute
 end
 
+desc 'Run Performance setup and a very short iteration of gatling tests locally'
```
Comment (er0ck) [inline rakefile]: this looks like another opportunity to split out via separate commit

File: rakefile (hunk 2 above)
Comment (samwoods1) [reply rakefile]: Oh, the check for BEAKER_INSTALL_TYPE makes sense because some of the other env vars are optional depending on what it is set to, but you're right, there's no need for this check.

File: rakefile (hunk 3 above)
Comment (samwoods1) [reply rakefile]: Actually, I need to update my PR, apparently this is not going to be respected either way with the new ABS workflow, so we'll need to do our own cleanup with the new de-provision task.

File: rakefile (hunk 2 above)
Comment (er0ck) [reply rakefile]: yeah, there is a ticket for pattern matching on env var values for tiller.  but i think that thing needs empty env var and other edge cases of the env var interface to be cleared up first anyway

File: rakefile (hunk 4 above)
Comment (samwoods1) [reply rakefile]: Hmm, looks like I copied that task and was going to modify it, but ended up making things work a different way, I'm removing this.

File: rakefile [hunk 5]
```diff
@@ -1,13 +1,85 @@
...
+        sleep(5)
+      end
+    end
+  end
+
+end
+
+desc 'Execute beaker performance setup and tests against existing hosts specified by ABS_RESOURCE_HOSTS env var, intended to be used in CI - use task "performance" for local/dev execution'
+rototiller_task :performance_without_provision do |t|
+  raise "The ABS_RESOURCE_HOSTS env var must be set to a valid value to continue" unless
```
Comment (RandellP) [inline rakefile]: seems odd that the error message mentions only one env variable, but the condition checks for either of two env variables.

File: setup/helpers/perf_helper.rb [hunk 6]
```diff
@@ -343,19 +249,33 @@ def perf_install_foss
...
+        master.upgrade_package(nss_package_name)
+      else
+        Beaker::Log.warn("Don't know what nss package to use for #{variant} so not installing one")
+      end
     end
 
-    step "Sign agent certificates" do
-      # Failed agent run will request a certificate
-      on agents, 'puppet agent -t', {:accept_all_exit_codes => true}
-      on master, 'puppet cert sign --all'
```
Comment (RandellP) [inline setup/helpers/perf_helper.rb]: where did the cert signing go?  I don't see anything signing certs to make up for this line being removed.

File: rakefile [hunk 7]
```diff
@@ -1,13 +1,85 @@
...
+        puts ex
+        puts ex.backtrace
+      end
+      if res.nil? || res.body == ''
+        puts 'Awaiting ABS_RESOURCE_HOSTS response from ABS...'
+        sleep(10)
+      end
+    end
+  end
+  ENV['ABS_RESOURCE_HOSTS'] = res.body
```
Comment (er0ck) [inline rakefile]: do we really want to be writing to the user's env var space? ever?

Comment (samwoods1) [reply rakefile]: ABS requires this env var to be set... What would the alternative be?

File: setup/helpers/perf_helper.rb (hunk 6 above)
Comment (samwoods1) [reply setup/helpers/perf_helper.rb]: It is part of the normal setup. I had to call it explicitly before because I couldn't use the default setup method, but now that we have access to internal resources it uses the default setup which takes care of it.

File: rakefile (hunk 7 above)
Comment (er0ck) [reply rakefile]: it does?  i thought abs was a remote webservice.

Comment (samwoods1) [reply rakefile]: More specifically, beaker-abs requires it

Comment (er0ck) [reply rakefile]: ah.  right.  that's annoying, but difficult to provide a better interface i guess.  
i'm glad you are mentioning that you are setting it on the next line, to the user.

carry on

File: rakefile [hunk 8]
```diff
@@ -1,13 +1,79 @@
...
 
 task :default => :performance
 
-desc 'Run Performance setup'
-rototiller_task :performance do |t|
+desc 'Provision systems using ABS (or use already set ABS_RESOURCE_HOSTS) and then execute beaker performance setup and tests'
+rototiller_task :performance do
+  Rake::Task["performance_provision_with_abs"].execute unless ENV['ABS_RESOURCE_HOSTS']
+  Rake::Task["performance_without_provision"].execute
+  Rake::Task["performance_deprovision_with_abs"].execute unless ENV['BEAKER_PRESERVE_HOSTS'] == 'always' || (@beaker_cmd.result.exit_code != 0 && ENV['BEAKER_PRESERVE_HOSTS'] == 'onfail')
```
Comment (er0ck) [inline rakefile]: what happens when @beaker_cmd has not been set?

File: setup/helpers/perf_helper.rb [hunk 9]
```diff
@@ -343,19 +249,33 @@ def perf_install_foss
       if package_build_version
         Beaker::Log.notify("Installing OSS Puppet Server version '#{package_build_version}'")
         install_puppetlabs_dev_repo master, 'puppetserver', package_build_version,
-                                    repo_config_dir, @install_opts
-      else
+                                    repo_config_dir, install_opts
+        else
```
Comment (er0ck) [inline setup/helpers/perf_helper.rb]: errant indent?

File: rakefile (hunk 8 above)
Comment (samwoods1) [reply rakefile]: Hmm, looks like this would throw an exception and never run, which probably would be the desired behavior. We probably would want to assume a failure if @beaker_cmd has not been set.

Comment (er0ck) [reply rakefile]: i guess that's fine.  do you think the user will know why or what task to run first?

Comment (samwoods1) [reply rakefile]: I tried to update the readme also. I am also updating the descriptions of the tasks to try to be more clear.

**Comment by samwoods1** [comment]
I added a new commit to this PR to support running against VMPooler for the acceptance tests.

File: rakefile [hunk 10]
```diff
@@ -1,13 +1,83 @@
...
 
 task :default => :performance
 
-desc 'Run Performance setup'
-rototiller_task :performance do |t|
+desc 'Provision systems using ABS (or use already set ABS_RESOURCE_HOSTS) and then execute beaker performance setup and tests'
+rototiller_task :performance do
+  Rake::Task["performance_provision_with_abs"].execute unless ENV['ABS_RESOURCE_HOSTS']
+  Rake::Task["performance_without_provision"].execute
+  Rake::Task["performance_deprovision_with_abs"].execute unless ENV['BEAKER_PRESERVE_HOSTS'] == 'always' || ((@beaker_cmd.nil? || @beaker_cmd.result.exit_code != 0) && ENV['BEAKER_PRESERVE_HOSTS'] == 'onfail')
```
Comment (RandellP) [inline rakefile]: This looks fine, but is pretty hard to read (mostly due to it's length and where line breaks might end up).  Not worth a new commit to fix, but if you end up making a commit to this PR for another reason, it would be nice to make this more readable in that commit.

File: rakefile [hunk 11]
```diff
@@ -1,13 +1,83 @@
...
+  res
+end
+
+desc 'Provision systems for the performance task with ABS - included in the performance task'
+rototiller_task :performance_provision_with_abs do
+  # Creating a random job id, since this is required by ABS and needs to be unique. In CI, this task will not be executed.
+  ENV['ABS_JOB_ID'] = rand(100000..999999).to_s unless ENV['ABS_JOB_ID']
+  uri = URI("#{BASE_ABS_URI}request")
+
+  # Allows us to switch between AWS and VMPooler
```
Comment (RandellP) [inline rakefile]: how?

File: rakefile [hunk 12]
```diff
@@ -104,7 +174,7 @@ rototiller_task :performance do |t|
     end
     command.add_option do |option|
       option.name = '--preserve-hosts'
-      option.message = 'Whether to preserve hosts or not'
+      option.message = 'Whether to preserve hosts or not, if using the default ABS workflow, this will not be respected.'
```
Comment (RandellP) [inline rakefile]: Thanks to the great code you added, this is no accepted... isn't it?

File: rakefile [hunk 13]
```diff
@@ -143,6 +214,8 @@ desc 'Run Performance setup and a very short iteration of gatling tests'
 rototiller_task :acceptance do |t|
   ENV['ENVIRONMENT_TYPE'] = 'gatling'
   ENV['PUPPET_GATLING_SCENARIO'] = 'acceptance.json'
+  ENV['BEAKER_HOSTS'] = 'config/beaker_hosts/pe-vmpooler.cfg'
+  ENV['ABS_OS'] = 'centos-7-x86_64'
```
Comment (RandellP) [inline rakefile]: on line 55 you set this if not set... why does it need setting here?  And if here, why on 55?  And why different values.  It may all be good, and if so, some more comments would really help.

**Comment by RandellP** [comment]
I thought we were concerned about always running acceptance in vmpooler, since aws has lots of it's own challenges?   I think we could create two tasks acceptance_abs and acceptance_vmpooler.  Then have acceptance run the vmpooler one.  That way we can run vmpooler for fast feedback, and then use abs as a final check before merge.

Also, the readme needs an update mentioning how to make a run happen in vmpooler one way or another.  I think the section "Set up Puppet Enterprise and Gatling but do not execute a gatling scenario" is where it is most useful.  You wouldn't want to do a full run and execute in vmpooler... but a no execute run would be good for dev work.

**Comment by samwoods1** [comment]
@RandellP I think the vast majority of the time we are just going to want to run the acceptance tests in vmpooler. With the new AWS account, and the fact that we're using the base Centos7 image in both AWS and VMPooler now, things should be much more stable.  If we really need to run acceptance tests in AWS instead of vmpooler, we can kick it off by simply changing some env vars before the run.

**Comment by RandellP** [comment]
tests passed.  The readme could use a blurb about how to run acceptance in aws.  Just document the env vars so a person doesn't have to go digging to find out how to do it.

File: rakefile [hunk 14]
```diff
@@ -1,13 +1,85 @@
...
+      else
+        break
+      end
+    end
+  end
+  res
+end
+
+desc 'Provision systems for the performance task with ABS - included in the performance task'
+rototiller_task :performance_provision_with_abs do
```
Comment (er0ck) [inline rakefile]: these tasks should probably be organized by namespace later...
something like `performance:provision_abs` etc

but i'm fine deferring UX changes like this until later

File: rakefile [hunk 15]
```diff
@@ -1,13 +1,85 @@
...
+  # Creating a random job id, since this is required by ABS and needs to be unique. In CI, this task will not be executed.
+  ENV['ABS_JOB_ID'] = rand(100000..999999).to_s unless ENV['ABS_JOB_ID']
+  uri = URI("#{BASE_ABS_URI}request")
```
Comment (er0ck) [inline rakefile]: does this have a ticket for fixing up later?  we could have easily used a truly unique id 🤷‍♂️ (uuid)

File: rakefile [hunk 16]
```diff
@@ -1,13 +1,85 @@
...
+
+  # Allows us to switch between AWS and VMPooler by selecting different ABS os's.
+  # centos-7-x86-64-west is an AWS image, centos-7-x86_64 is vmpooler
+  abs_os = ENV['ABS_OS'].nil? ? 'centos-7-x86-64-west' : ENV['ABS_OS']
```
Comment (er0ck) [inline rakefile]: why isn't this a rototiller env_var?

File: setup/helpers/perf_helper.rb [hunk 17]
```diff
@@ -668,7 +572,9 @@ def setup_gatling_proxy
     end
     step 'install rvm, bundler' do
       on metric, 'gpg --keyserver hkp://keys.gnupg.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3'
-      on metric, 'curl -sSL https://get.rvm.io | bash -s stable --ruby=2.2.5'
+      on metric, 'curl -sSL https://get.rvm.io | bash -s stable'
+      # Work around install issue in Centos7: https://github.com/rvm/rvm/issues/4234
+      on metric, 'rvm install 2.2.5 --disable-binary'
```
Comment (er0ck) [inline setup/helpers/perf_helper.rb]: if this is to work around known 2.2.5 rvm issues in centos, this change should be in a separate commit.  oh, i see the comment now 🤣 
but yeah, separate commit for ease of revert later

Comment (er0ck) [reply setup/helpers/perf_helper.rb]: i guess install also implicitly uses that latest version?  or maybe if there is only one version?
the man page is vague on it.

File: rakefile (hunk 15 above)
Comment (samwoods1) [reply rakefile]: It expects an integer... So, no uuid.

**Comment by er0ck** [comment]
:shipit:
